## to do:
- work out how to show number of visits for a plate
- sort plates by number of visits (ideally N > 20)
- figure out how to turn data on this plate into pandas df
- run pca to analyse the N RV measurements of those ~200 stars together

## questions:
- what does 'bestars' represent?
- what is observed spectrum template matching

In [152]:
import numpy as np
from matplotlib import pyplot as plt
from astropy.io import fits
import fitsio
import statistics as stat
import pandas as pd

## using fitsio
### figuring out plate id data

In [155]:
# Read in fits file data using fitsio

data = fitsio.read("allVisit-r12-l33.fits")

In [176]:
# Define variables for working with PLATE data

plates = data["PLATE"]
plateslist = list(plates) #[5:]) # only numerical plate ids

print(plates.size) # Same as total number of visits

In [177]:
# Strip whitespace in plates list

for i, s in enumerate(plateslist):
    plateslist[i] = s.strip()

print(plateslist[:7]) # Check if strip was successful

['Bestars', 'Bestars', 'calibration', 'calibration', 'calibration', '7545', '7545']


In [178]:
# Plate with greatest number of visits

stat.multimode(plateslist)

['9290']

In [179]:
# Make list of unique plates

unique_plates = []
total_up = 0

for i in plateslist:
    if i not in unique_plates:
        total_up = total_up + 1
        unique_plates.append(i)

print('Number of unique plates:', total_up)

Number of unique plates: 2383


In [182]:
# Count the times that each plate appears; i.e. visits per plate

visit_count = []
for u in unique_plates:
    nvisits = plateslist.count(u)
    visit_count.append(nvisits)

In [197]:
# Create pandas df for visits per plate

plates_mdlist = [unique_plates, visit_count]

df = pd.DataFrame(plates_mdlist).transpose()
df.columns = ['Plate ID', 'Visit Count']

print(df)

         Plate ID Visit Count
0         Bestars          48
1     calibration         237
2            7545         792
3            7917         500
4            5583         794
...           ...         ...
2378         9260         794
2379         5582         794
2380         7540         792
2381        11039         265
2382         8655        1056

[2383 rows x 2 columns]


## using astropy
### figuring out rv data
- Let's work with plate '7545', which has 792 visits
- Want to sort OBSVHELIO for plate '7545' to get RV
- Want to sort MJD for plate '7545' to get time

In [202]:
# HDU stuff

hdulist = fits.open('allVisit-r12-l33.fits')

hdulist.info()

header = hdulist[1].header
data = hdulist[1].data

hdulist.close()

In [219]:
# Set up data parameters

allplates = list(data['PLATE'])
allmjd = list(data['MJD'])
allrvs = list(data['OBSVHELIO']) # Heliocentric relative RV from 'observed spectrum template matching'

In [234]:
# Create pandas df

datalst = [allplates, allmjd, allrvs]

df = pd.DataFrame(datalst).transpose()
df.columns = ['Plate ID', 'MJD', 'OBSVHELIO (km/s)']

print(df)

             Plate ID    MJD OBSVHELIO (km/s)
0             Bestars  58017         -523.166
1             Bestars  58022         -526.084
2         calibration  56398          25.5533
3         calibration  56778          10.5618
4         calibration  57743         -36.2044
...               ...    ...              ...
1778787          5583  56261         -60.9357
1778788          5583  56284          -60.901
1778789          6560  56584         -5.20138
1778790          6560  56588         -5.03455
1778791          6560  56613         -5.51745

[1778792 rows x 3 columns]


In [232]:
# Show data only if plate ID is '7545'

dff = df[(df['Plate ID'] == 7545)]

In [233]:
print(dff)

Empty DataFrame
Columns: [Plate ID, MJD, OBS V_HELIO (km/s)]
Index: []
